File: android/app/src/main/java/com/alphaauxiliary/fitgenerator/map/CoordinateTransforms.kt
package com.alphaauxiliary.fitgenerator.map

import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.asin
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.sqrt

internal enum class CoordinateSystem {
    WGS84,
    GCJ02,
    BD09,
}

internal data class GeoCoordinate(
    val latitude: Double,
    val longitude: Double,
)

internal object CoordinateTransforms {
    private const val SEMI_MAJOR_AXIS = 6_378_245.0
    private const val ECCENTRICITY_SQUARED = 0.00669342162296594323
    private const val BD_PI = PI * 3_000.0 / 180.0
    private const val EARTH_RADIUS_METERS = 6_371_008.8
    private const val CHINA_MINIMUM_LATITUDE = 0.8293
    private const val CHINA_MAXIMUM_LATITUDE = 55.8271
    private const val CHINA_MINIMUM_LONGITUDE = 72.004
    private const val CHINA_MAXIMUM_LONGITUDE = 137.8347
    private const val INVERSE_TOLERANCE_DEGREES = 1e-7
    private const val MAXIMUM_INVERSE_ITERATIONS = 10

    fun toWgs84(point: GeoCoordinate, source: CoordinateSystem): GeoCoordinate {
        validateCoordinate(point)
        return when (source) {
            CoordinateSystem.WGS84 -> point
            CoordinateSystem.GCJ02 -> gcj02ToWgs84(point)
            CoordinateSystem.BD09 -> gcj02ToWgs84(bd09ToGcj02(point))
        }
    }

    fun fromWgs84(point: GeoCoordinate, target: CoordinateSystem): GeoCoordinate {
        validateCoordinate(point)
        return when (target) {
            CoordinateSystem.WGS84 -> point
            CoordinateSystem.GCJ02 -> wgs84ToGcj02(point)
            CoordinateSystem.BD09 -> gcj02ToBd09(wgs84ToGcj02(point))
        }
    }

    fun wgs84ToGcj02(point: GeoCoordinate): GeoCoordinate {
        validateCoordinate(point)
        if (isOutsideChina(point)) {
            return point
        }

        var latitudeOffset = transformLatitude(
            longitude = point.longitude - 105.0,
            latitude = point.latitude - 35.0,
        )
        var longitudeOffset = transformLongitude(
            longitude = point.longitude - 105.0,
            latitude = point.latitude - 35.0,
        )
        val latitudeRadians = degreesToRadians(point.latitude)
        val sineLatitude = sin(latitudeRadians)
        val magic = 1.0 - ECCENTRICITY_SQUARED * sineLatitude * sineLatitude
        val squareRootMagic = sqrt(magic)
        latitudeOffset = latitudeOffset * 180.0 /
            (
                (SEMI_MAJOR_AXIS * (1.0 - ECCENTRICITY_SQUARED)) /
                    (magic * squareRootMagic) * PI
                )
        longitudeOffset = longitudeOffset * 180.0 /
            (SEMI_MAJOR_AXIS / squareRootMagic * cos(latitudeRadians) * PI)

        return GeoCoordinate(
            latitude = point.latitude + latitudeOffset,
            longitude = point.longitude + longitudeOffset,
        )
    }

    fun gcj02ToWgs84(point: GeoCoordinate): GeoCoordinate {
        validateCoordinate(point)
        if (isOutsideChina(point)) {
            return point
        }

        var estimate = point
        repeat(MAXIMUM_INVERSE_ITERATIONS) {
            val transformed = wgs84ToGcj02(estimate)
            val latitudeError = transformed.latitude - point.latitude
            val longitudeError = transformed.longitude - point.longitude
            estimate = GeoCoordinate(
                latitude = estimate.latitude - latitudeError,
                longitude = estimate.longitude - longitudeError,
            )

            if (
                abs(latitudeError) < INVERSE_TOLERANCE_DEGREES &&
                abs(longitudeError) < INVERSE_TOLERANCE_DEGREES
            ) {
                return estimate
            }
        }
        return estimate
    }

    fun gcj02ToBd09(point: GeoCoordinate): GeoCoordinate {
        validateCoordinate(point)
        val radius = sqrt(
            point.longitude * point.longitude + point.latitude * point.latitude,
        ) + 0.00002 * sin(point.latitude * BD_PI)
        val angle = atan2(point.latitude, point.longitude) +
            0.000003 * cos(point.longitude * BD_PI)

        return GeoCoordinate(
            latitude = radius * sin(angle) + 0.006,
            longitude = radius * cos(angle) + 0.0065,
        )
    }

    fun bd09ToGcj02(point: GeoCoordinate): GeoCoordinate {
        validateCoordinate(point)
        val longitude = point.longitude - 0.0065
        val latitude = point.latitude - 0.006
        val radius = sqrt(longitude * longitude + latitude * latitude) -
            0.00002 * sin(latitude * BD_PI)
        val angle = atan2(latitude, longitude) -
            0.000003 * cos(longitude * BD_PI)

        return GeoCoordinate(
            latitude = radius * sin(angle),
            longitude = radius * cos(angle),
        )
    }

    fun distanceMeters(first: GeoCoordinate, second: GeoCoordinate): Double {
        validateCoordinate(first)
        validateCoordinate(second)
        val latitudeDelta = degreesToRadians(second.latitude - first.latitude)
        val longitudeDelta = degreesToRadians(second.longitude - first.longitude)
        val firstLatitude = degreesToRadians(first.latitude)
        val secondLatitude = degreesToRadians(second.latitude)
        val haversine = sin(latitudeDelta / 2.0) * sin(latitudeDelta / 2.0) +
            cos(firstLatitude) * cos(secondLatitude) *
            sin(longitudeDelta / 2.0) * sin(longitudeDelta / 2.0)
        val centralAngle = 2.0 * asin(sqrt(min(1.0, max(0.0, haversine))))
        return EARTH_RADIUS_METERS * centralAngle
    }

    private fun isOutsideChina(point: GeoCoordinate): Boolean =
        point.longitude < CHINA_MINIMUM_LONGITUDE ||
            point.longitude > CHINA_MAXIMUM_LONGITUDE ||
            point.latitude < CHINA_MINIMUM_LATITUDE ||
            point.latitude > CHINA_MAXIMUM_LATITUDE

    private fun validateCoordinate(point: GeoCoordinate) {
        require(point.latitude.isFinite() && point.longitude.isFinite()) {
            "坐标必须是有限数值"
        }
        require(point.latitude in -90.0..90.0 && point.longitude in -180.0..180.0) {
            "坐标超出有效经纬度范围"
        }
    }

    private fun transformLatitude(longitude: Double, latitude: Double): Double {
        var transformed = -100.0 +
            2.0 * longitude +
            3.0 * latitude +
            0.2 * latitude * latitude +
            0.1 * longitude * latitude +
            0.2 * sqrt(abs(longitude))
        transformed += (
            20.0 * sin(6.0 * longitude * PI) +
                20.0 * sin(2.0 * longitude * PI)
            ) * 2.0 / 3.0
        transformed += (
            20.0 * sin(latitude * PI) +
                40.0 * sin(latitude / 3.0 * PI)
            ) * 2.0 / 3.0
        transformed += (
            160.0 * sin(latitude / 12.0 * PI) +
                320.0 * sin(latitude * PI / 30.0)
            ) * 2.0 / 3.0
        return transformed
    }

    private fun transformLongitude(longitude: Double, latitude: Double): Double {
        var transformed = 300.0 +
            longitude +
            2.0 * latitude +
            0.1 * longitude * longitude +
            0.1 * longitude * latitude +
            0.1 * sqrt(abs(longitude))
        transformed += (
            20.0 * sin(6.0 * longitude * PI) +
                20.0 * sin(2.0 * longitude * PI)
            ) * 2.0 / 3.0
        transformed += (
            20.0 * sin(longitude * PI) +
                40.0 * sin(longitude / 3.0 * PI)
            ) * 2.0 / 3.0
        transformed += (
            150.0 * sin(longitude / 12.0 * PI) +
                300.0 * sin(longitude / 30.0 * PI)
            ) * 2.0 / 3.0
        return transformed
    }

    private fun degreesToRadians(value: Double): Double = value * PI / 180.0
}

File: native-windows/FitGenerator.Native/Map/CoordinateTransforms.cs
namespace FitGenerator.Native.Map;

internal static class CoordinateTransforms
{
    private const double Pi = Math.PI;
    private const double SemiMajorAxis = 6_378_245.0;
    private const double EccentricitySquared = 0.00669342162296594323;
    private const double BdPi = Pi * 3_000.0 / 180.0;
    private const double EarthRadiusMeters = 6_371_008.8;
    private const double ChinaMinimumLatitude = 0.8293;
    private const double ChinaMaximumLatitude = 55.8271;
    private const double ChinaMinimumLongitude = 72.004;
    private const double ChinaMaximumLongitude = 137.8347;
    private const double InverseToleranceDegrees = 1e-7;
    private const int MaximumInverseIterations = 10;

    internal static GeoCoordinate ToWgs84(GeoCoordinate value, CoordinateSystem source)
    {
        ValidateCoordinate(value, nameof(value));
        return source switch
        {
            CoordinateSystem.Wgs84 => value,
            CoordinateSystem.Gcj02 => Gcj02ToWgs84(value),
            CoordinateSystem.Bd09 => Gcj02ToWgs84(Bd09ToGcj02(value)),
            _ => throw new ArgumentOutOfRangeException(nameof(source), source, "不支持的坐标系"),
        };
    }

    internal static GeoCoordinate FromWgs84(GeoCoordinate value, CoordinateSystem target)
    {
        ValidateCoordinate(value, nameof(value));
        return target switch
        {
            CoordinateSystem.Wgs84 => value,
            CoordinateSystem.Gcj02 => Wgs84ToGcj02(value),
            CoordinateSystem.Bd09 => Gcj02ToBd09(Wgs84ToGcj02(value)),
            _ => throw new ArgumentOutOfRangeException(nameof(target), target, "不支持的坐标系"),
        };
    }

    internal static GeoCoordinate Wgs84ToGcj02(GeoCoordinate value)
    {
        ValidateCoordinate(value, nameof(value));
        if (IsOutsideChina(value))
        {
            return value;
        }

        var latitudeOffset = TransformLatitude(
            value.Longitude - 105.0,
            value.Latitude - 35.0);
        var longitudeOffset = TransformLongitude(
            value.Longitude - 105.0,
            value.Latitude - 35.0);
        var latitudeRadians = DegreesToRadians(value.Latitude);
        var sineLatitude = Math.Sin(latitudeRadians);
        var magic = 1.0 - EccentricitySquared * sineLatitude * sineLatitude;
        var squareRootMagic = Math.Sqrt(magic);
        latitudeOffset = latitudeOffset * 180.0
            / ((SemiMajorAxis * (1.0 - EccentricitySquared))
                / (magic * squareRootMagic)
                * Pi);
        longitudeOffset = longitudeOffset * 180.0
            / (SemiMajorAxis / squareRootMagic * Math.Cos(latitudeRadians) * Pi);

        return new GeoCoordinate(
            value.Latitude + latitudeOffset,
            value.Longitude + longitudeOffset);
    }

    internal static GeoCoordinate Gcj02ToWgs84(GeoCoordinate value)
    {
        ValidateCoordinate(value, nameof(value));
        if (IsOutsideChina(value))
        {
            return value;
        }

        var estimate = value;
        for (var iteration = 0; iteration < MaximumInverseIterations; iteration++)
        {
            var transformed = Wgs84ToGcj02(estimate);
            var latitudeError = transformed.Latitude - value.Latitude;
            var longitudeError = transformed.Longitude - value.Longitude;
            estimate = new GeoCoordinate(
                estimate.Latitude - latitudeError,
                estimate.Longitude - longitudeError);

            if (Math.Abs(latitudeError) < InverseToleranceDegrees
                && Math.Abs(longitudeError) < InverseToleranceDegrees)
            {
                break;
            }
        }

        return estimate;
    }

    internal static GeoCoordinate Gcj02ToBd09(GeoCoordinate value)
    {
        ValidateCoordinate(value, nameof(value));
        var radius = Math.Sqrt(
                value.Longitude * value.Longitude
                + value.Latitude * value.Latitude)
            + 0.00002 * Math.Sin(value.Latitude * BdPi);
        var angle = Math.Atan2(value.Latitude, value.Longitude)
            + 0.000003 * Math.Cos(value.Longitude * BdPi);

        return new GeoCoordinate(
            radius * Math.Sin(angle) + 0.006,
            radius * Math.Cos(angle) + 0.0065);
    }

    internal static GeoCoordinate Bd09ToGcj02(GeoCoordinate value)
    {
        ValidateCoordinate(value, nameof(value));
        var longitude = value.Longitude - 0.0065;
        var latitude = value.Latitude - 0.006;
        var radius = Math.Sqrt(longitude * longitude + latitude * latitude)
            - 0.00002 * Math.Sin(latitude * BdPi);
        var angle = Math.Atan2(latitude, longitude)
            - 0.000003 * Math.Cos(longitude * BdPi);

        return new GeoCoordinate(
            radius * Math.Sin(angle),
            radius * Math.Cos(angle));
    }

    internal static double DistanceMeters(GeoCoordinate first, GeoCoordinate second)
    {
        ValidateCoordinate(first, nameof(first));
        ValidateCoordinate(second, nameof(second));
        var latitudeDelta = DegreesToRadians(second.Latitude - first.Latitude);
        var longitudeDelta = DegreesToRadians(second.Longitude - first.Longitude);
        var firstLatitude = DegreesToRadians(first.Latitude);
        var secondLatitude = DegreesToRadians(second.Latitude);
        var haversine = Math.Sin(latitudeDelta / 2.0) * Math.Sin(latitudeDelta / 2.0)
            + Math.Cos(firstLatitude)
            * Math.Cos(secondLatitude)
            * Math.Sin(longitudeDelta / 2.0)
            * Math.Sin(longitudeDelta / 2.0);
        var centralAngle = 2.0 * Math.Asin(Math.Sqrt(Math.Clamp(haversine, 0.0, 1.0)));
        return EarthRadiusMeters * centralAngle;
    }

    private static bool IsOutsideChina(GeoCoordinate value) =>
        value.Longitude < ChinaMinimumLongitude
        || value.Longitude > ChinaMaximumLongitude
        || value.Latitude < ChinaMinimumLatitude
        || value.Latitude > ChinaMaximumLatitude;

    private static void ValidateCoordinate(GeoCoordinate value, string parameterName)
    {
        if (!double.IsFinite(value.Latitude) || !double.IsFinite(value.Longitude))
        {
            throw new ArgumentOutOfRangeException(parameterName, value, "坐标必须是有限数值");
        }

        if (value.Latitude is < -90.0 or > 90.0
            || value.Longitude is < -180.0 or > 180.0)
        {
            throw new ArgumentOutOfRangeException(parameterName, value, "坐标超出有效经纬度范围");
        }
    }

    private static double TransformLatitude(double longitude, double latitude)
    {
        var transformed = -100.0
            + 2.0 * longitude
            + 3.0 * latitude
            + 0.2 * latitude * latitude
            + 0.1 * longitude * latitude
            + 0.2 * Math.Sqrt(Math.Abs(longitude));
        transformed += (20.0 * Math.Sin(6.0 * longitude * Pi)
                + 20.0 * Math.Sin(2.0 * longitude * Pi))
            * 2.0
            / 3.0;
        transformed += (20.0 * Math.Sin(latitude * Pi)
                + 40.0 * Math.Sin(latitude / 3.0 * Pi))
            * 2.0
            / 3.0;
        transformed += (160.0 * Math.Sin(latitude / 12.0 * Pi)
                + 320.0 * Math.Sin(latitude * Pi / 30.0))
            * 2.0
            / 3.0;
        return transformed;
    }

    private static double TransformLongitude(double longitude, double latitude)
    {
        var transformed = 300.0
            + longitude
            + 2.0 * latitude
            + 0.1 * longitude * longitude
            + 0.1 * longitude * latitude
            + 0.1 * Math.Sqrt(Math.Abs(longitude));
        transformed += (20.0 * Math.Sin(6.0 * longitude * Pi)
                + 20.0 * Math.Sin(2.0 * longitude * Pi))
            * 2.0
            / 3.0;
        transformed += (20.0 * Math.Sin(longitude * Pi)
                + 40.0 * Math.Sin(longitude / 3.0 * Pi))
            * 2.0
            / 3.0;
        transformed += (150.0 * Math.Sin(longitude / 12.0 * Pi)
                + 300.0 * Math.Sin(longitude / 30.0 * Pi))
            * 2.0
            / 3.0;
        return transformed;
    }

    private static double DegreesToRadians(double value) => value * Pi / 180.0;
}
